sdk/src/main/scala/io/horizen/account/api/rpc/service/Backend.scala

```diff
@@ -57,49 +60,100 @@ object Backend {
-    val blocks: Integer = (blockCount * 2).min(blockHeight).min(1024)
-
-    // define limit for included gas prices each block
-    val limit = 3
-    val prices: Seq[BigInteger] = {
-      var collected = 0
-      var moreBlocksNeeded = false
-      // Return lowest tx gas prices of each requested block, sorted in ascending order.
-      // Queries up to 2*blockCount blocks, but stops in range > blockCount if enough samples were found.
-      (0 until blocks).withFilter(_ => !moreBlocksNeeded || collected < 2).map { i =>
-        val block = history
-          .blockIdByHeight(blockHeight - i)
-          .map(ModifierId(_))
-          .flatMap(history.getStorageBlockById)
-          .get
-        val blockPrices = getBlockPrices(block, ignorePrice, limit)
-        collected += blockPrices.length
-        if (i >= blockCount) moreBlocksNeeded = true
-        blockPrices
+    var number = history.getCurrentHeight
+    val headHash = history.bestBlockId
+    val (lastHead : Option[ModifierId], lastPrice: BigInteger) = tipCache.fold(
+      (Option.empty[ModifierId], BigInteger.ZERO)
+    )(cache =>
+      (Some(cache.blockHash), cache.value)
+    )
+
+    // If the latest gasprice is still valid, return it.
+    if (lastHead.isDefined && lastHead.get == headHash){
+      return lastPrice
+    }
+
+    var sent = 0
+    var exp = 0
+    var prices: Seq[Option[Seq[BigInteger]]] =  Seq() //in go-ethereum this is called result
+    var results: Seq[BigInteger] = Seq()
+    while (sent < blockCount && number > 0){
+      prices = prices :+ getBlockPrices(history, number, ignorePrice, SUGGEST_TIP_TX_LIMIT)
+      sent += 1
+      exp += 1
+      number -= 1
+    }
+    while (exp > 0){
+      val pricesOfABlock : Option[Seq[BigInteger]] = prices.head
+      prices = prices.drop(1)
+      if (pricesOfABlock.isEmpty) {
+        //if we are here we had some errors collecting the blocks
+        log.warn("Error retrieving blocks in history while suggesting tip cap, returning the cached one")
+        return lastPrice
+      }
+      exp -= 1
+      var res : Seq[BigInteger] = pricesOfABlock.get
+      if (res.isEmpty) {
+        // Nothing returned. There are two special cases here:
+        // - The block is empty
+        // - All the transactions included are sent by the miner itself.
+        // In these cases, use the latest calculated price for sampling.
+        res = Seq(lastPrice)
       }
-    }.flatten
 
-    prices
-      .sorted
-      .lift((prices.length - 1) * percentile / 100)
-      .getOrElse(BigInteger.ZERO)
-      .min(maxPrice)
+      // Besides, in order to collect enough data for sampling, if nothing
+      // meaningful returned, try to query more blocks. But the maximum
+      // is 2*checkBlocks.
+      if (res.length == 1 && results.length + 1 + exp < blockCount * 2 && number > 0) {
+        prices = prices :+ getBlockPrices(history, number, ignorePrice, 3)
```

Comment:
This 3 should still refer to `SUGGEST_TIP_TX_LIMIT`, right? If yes, we can replace it

Reply:
yes correct, thanks!